### PR TITLE
Update time function model 

### PR DESCRIPTION
**Pull request by thalestas**
This PR updates the time function model as requested and add a regression test for it.

### CODE REVIEW

File: src/c2goto/library/time.c [hunk 1]
```diff
@@ -6,10 +6,12 @@
 #include <time.h>
 #undef time
 
+time_t __VERIFIER_nondet_time_t();
```
Comment (lucasccordeiro) [inline src/c2goto/library/time.c:9]: Could you please add the declaration of this function in:

- https://github.com/esbmc/esbmc/blob/master/src/clang-c-frontend/clang_c_language.cpp#L319
- https://github.com/esbmc/esbmc/blob/master/src/ansi-c/ansi_c_language.cpp#L75

File: regression/esbmc/time/test.desc [hunk 2]
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+
+^VERIFICATION SUCCESSFUL$
```
Comment (lucasccordeiro) [inline regression/esbmc/time/test.desc]: Could you also add a test case that should fail?

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]
@thalestas: many thanks for creating this PR. 

Could you please address two minor comments that I posted here?

File: src/c2goto/library/time.c [hunk 3]
```diff
@@ -6,10 +6,12 @@
...
+
 time_t time(time_t *tloc)
 {
-  time_t res;
-  if(!tloc)
+  time_t res = __VERIFIER_nondet_time_t();
```
Comment (mikhailramalho) [inline src/c2goto/library/time.c:14]: Please, add the ESBMC_HIDE:; macro here.

File: src/c2goto/library/time.c [hunk 4]
```diff
@@ -6,10 +6,12 @@
...
+  if(tloc != NULL)
```
Comment (mikhailramalho) [inline src/c2goto/library/time.c]: No need to compare against NULL (btw, is NULL defined in this header)? `if(tloc)` should be enough.

**Review by mikhailramalho — CHANGES_REQUESTED** [review]

File: regression/esbmc/time_02/test.desc [hunk 5]
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+
+^VERIFICATION FAILED$
```
Comment (mikhailramalho) [inline regression/esbmc/time_02/test.desc:4]: Why is this program failing?

File: src/ansi-c/ansi_c_language.cpp [hunk 6]
```diff
@@ -85,6 +87,7 @@ static void internal_additions(std::string &code)
     "_Bool __VERIFIER_nondet_bool();\n"
     "float __VERIFIER_nondet_float();\n"
     "double __VERIFIER_nondet_double();\n"
+    "time_t __VERIFIER_nondet_time_t();\n"
```
Comment (mikhailramalho) [inline src/ansi-c/ansi_c_language.cpp:90]: No need to define the `__VERIFIER_nondet_time_t()` function as it is not part of the TACAS rules.

File: src/c2goto/library/time.c [hunk 7]
```diff
@@ -10,8 +10,9 @@ time_t __VERIFIER_nondet_time_t();
 
 time_t time(time_t *tloc)
 {
+__ESBMC_HIDE:;
   time_t res = __VERIFIER_nondet_time_t();
```
Comment (mikhailramalho) [inline src/c2goto/library/time.c:14]: You can replace it with `nondet_time_t()`.

File: src/clang-c-frontend/clang_c_language.cpp [hunk 8]
```diff
@@ -262,6 +262,7 @@ std::string clang_c_languaget::internal_additions()
   std::string intrinsics =
     R"(
 # 1 "esbmc_intrinsics.h" 1
+typedef long time_t;
```
Comment (mikhailramalho) [inline src/clang-c-frontend/clang_c_language.cpp]: I don't think this is correct, as it is platform dependent: https://en.cppreference.com/w/c/chrono/time_t

I suggest removing both `nondet_time_t()` and `__VERIFIER_nondet_time_t()` and leave it as it was before: just define it in the time.c model lib.

@lucasccordeiro, do you agree?

Comment (lucasccordeiro) [reply src/clang-c-frontend/clang_c_language.cpp]: I agree with you.

I had previously suggested adding the prototype of this function in:

- https://github.com/esbmc/esbmc/blob/master/src/clang-c-frontend/clang_c_language.cpp#L319
- https://github.com/esbmc/esbmc/blob/master/src/ansi-c/ansi_c_language.cpp#L75

Thales told me that he had compilation problems. However, if we look at the spec, we can see that it is platform dependent.

@thales: Could you please remove the declaration of these functions from the files above?

Many thanks,
Lucas

File: regression/esbmc/time_02/test.desc (hunk 5 above)
Comment (thalestas) [reply regression/esbmc/time_02/test.desc:4]: As time() function expects time_t pointer, if an int is passed it results in a out of bound.

Counterexample:

State 1 file time.c line 16 function time thread 0
----------------------------------------------------
Violated property:
  file time.c line 16 function time
  dereference failure: Access to object out of bounds

VERIFICATION FAILED

File: src/clang-c-frontend/clang_c_language.cpp (hunk 8 above)
Comment (thalestas) [reply src/clang-c-frontend/clang_c_language.cpp]: Sure!

File: src/clang-c-frontend/clang_c_language.cpp [hunk 9]
```diff
@@ -302,6 +302,7 @@ void *__ESBMC_memset(void *, int, unsigned int);
 void pthread_start_main_hook(void);
 void pthread_end_main_hook(void);
 
+
```
Comment (lucasccordeiro) [inline src/clang-c-frontend/clang_c_language.cpp]: Can you remove this line?

**Review by lucasccordeiro — APPROVED** [review]
Overall, it looks good to me. I have just one small comment.

File: src/clang-c-frontend/clang_c_language.cpp (hunk 9 above)
Comment (thalestas) [reply src/clang-c-frontend/clang_c_language.cpp]: Ok!

**Review by mikhailramalho — APPROVED** [review]
LGTM

**Pull request merged**